Mathlib/Topology/UniformSpace/Pi.lean

```diff
@@ -127,3 +127,20 @@ instance Pi.separated [∀ i, SeparatedSpace (α i)] : SeparatedSpace (∀ i, α
 #align Pi.separated Pi.separated
 
 end
+
+theorem CompleteSpace.iInf {ι : Sort*} {X : Type*} {u : ι → UniformSpace X}
+    (hu : ∀ i, @CompleteSpace X (u i))
+    (ht₀ : ∃ i₀, @T0Space X (u i₀).toTopologicalSpace ∧
+      ∀ i, (u i).toTopologicalSpace ≤ (u i₀).toTopologicalSpace) :
+    @CompleteSpace X (⨅ i, u i) := by
+  refine @CompleteSpace.mk X ?_ fun {f} hf ↦ ?_
+  rcases ht₀ with ⟨i₀, hsep₀, hi₀⟩
+  have hf' : ∀ i, Cauchy (uniformSpace := u i) f := fun i ↦ hf.mono_uniformSpace (iInf_le _ _)
+  choose x hfx using fun i ↦ @CompleteSpace.complete _ (u i) (hu i) f (hf' i)
+  have hx : ∀ i, x i = x i₀ := fun i ↦ by
+    let _ := u i₀
+    have := hf.1
+    exact tendsto_nhds_unique ((hfx i).trans <| nhds_mono (hi₀ i)) (hfx i₀)
+  use x i₀
+  rw [UniformSpace.toTopologicalSpace_iInf, nhds_iInf]
+  exact le_iInf fun i ↦ (hx i).symm ▸ hfx i
```

Comment:
I thought it was a bit sad not to keep the original sketch of the proof, which is a bit more natural in my opinion (and makes it clear why we need some separation, to ensure that the diagonal is closed in the product). So I played a bit with it, and ended up with this version (for which you need to import `UniformEmbedding`):
```lean
-- move to top of file
open Function

-- move somewhere else
theorem Function.range_const_eq_diagonal {α β : Type*} [hβ : Nonempty β] :
    Set.range (const α) = {f : α → β | ∀ x y, f x = f y} := by
  refine Set.range_eq_iff _ _ |>.mpr ⟨fun _ _ _ ↦ rfl, fun f hf ↦ ?_⟩
  rcases isEmpty_or_nonempty α with h|⟨⟨a⟩⟩
  · exact hβ.elim fun b ↦ ⟨b, Subsingleton.elim _ _⟩
  · exact ⟨f a, funext fun x ↦ hf _ _⟩

/- An infimum of complete uniformities is complete, as long as the whole family is bounded by
some common T2 topology. -/
theorem CompleteSpace.iInf {ι : Type*} {X : Type*} {u : ι → UniformSpace X} {t : TopologicalSpace X}
    (hu : ∀ i, @CompleteSpace X (u i)) (ht : @T2Space X t)
    (hut : ∀ i, (u i).toTopologicalSpace ≤ t) :
    @CompleteSpace X (⨅ i, u i) := by
  -- We can assume `X` is nonempty.
  nontriviality X
  -- The diagonal map `(X, ⨅ i, u i) → Π i, (X, u i)` is a uniform embedding.
  have : @UniformInducing X (ι → X) (⨅ i, u i) (Pi.uniformSpace (U := u)) (const ι) := by
    simp_rw [uniformInducing_iff_uniformSpace, Pi.uniformSpace_eq,
      UniformSpace.comap_iInf, ← UniformSpace.comap_comap]
    exact iInf_congr fun _ ↦ congrFun uniformSpace_comap_id _
  -- Hence, it suffices to show that its range, the diagonal, is closed in `Π i, (X, u i)`.
  simp_rw [@completeSpace_iff_isComplete_range _ _ (_) (_) _ this, range_const_eq_diagonal,
    Set.setOf_forall]
  -- The separation of `t` ensures that this is the case in `Π i, (X, t)`, hence the result
  -- since the topology associated to each `u i` is finer than `t`.
  have : Pi.topologicalSpace (t₂ := fun i ↦ (u i).toTopologicalSpace) ≤
         Pi.topologicalSpace (t₂ := fun _ ↦ t) :=
    iInf_mono fun i ↦ induced_mono <| hut i
  refine IsClosed.isComplete <| .mono ?_ this
  exact isClosed_iInter fun i ↦ isClosed_iInter fun j ↦
    isClosed_eq (continuous_apply _) (continuous_apply _)
```

Note that I made the statement a bit stronger (the T2 topology which bounds everything doesn't have to come from a complete uniform structure) and probably easier to use, and even if you keep your original sketch of proof you can definitely do the same strengthening.

It looks a bit longer because of the comments, but adapting your proof to the same statement gives exactly the same length. As I said I find this one a bit more natural but this is probably a matter of taste.

Reply:
Thanks! I replaced my proof with yours. It changes `{ι : Sort*}` to `{ι : Type*}` but I don't think that we should care in this case.